bugsnag/src/main/java/com/bugsnag/callbacks/DeviceCallback.java

```diff
@@ -39,10 +48,24 @@ private static String lookupHostname() {
             return hostname;
         }
 
-        // Resort to dns hostname lookup
+        // Resort to dns hostname lookup.
+        // Look up the hostname in a daemon thread, with a timeout of HOSTNAME_LOOKUP_TIMEOUT.
+        // This is to prevent applications hanging waiting for the dns lookup to resolve
+        ExecutorService lookupService = Executors.newSingleThreadExecutor(new DaemonThreadFactory());
+        Future<String> future = lookupService.submit(new Callable<String>() {
+            @Override
+            public String call() throws UnknownHostException {
+                return InetAddress.getLocalHost().getHostName();
+            }
+        });
```

Comment:
```suggestion
        FutureTask<String> future = new FutureTask<String>(new Callable<String>() {
            @Override
            public String call() throws UnknownHostException {
                return InetAddress.getLocalHost().getHostName();
            }
        });
        Thread resolverThread = new Thread(future, "Hostname Resolver");
        resolverThread.setDaemon(true);
        resolverThread.start();
```

I'm not convinced that using an `ExecutorService` here is the correct approach, as without shutting it down it will leak both the `ExecutorService` and `Thread`. Using a plain `Thread` gives us the same functionality, but will terminate as soon as its job is complete and in this case it holds no external references so shouldn't cause leaks.